crates/driver/src/logic/competition/solution/mod.rs

```diff
@@ -107,6 +109,66 @@ impl Solution {
             .map(|(spender, amount)| eth::Allowance { spender, amount }.into())
             .sorted()
     }
+
+    /// Simulate this solution on the blockchain.
+    async fn simulate(
+        &self,
+        eth: &Ethereum,
+        simulator: &Simulator,
+        // TODO Remove the auction parameter in a follow-up
+        auction: &competition::Auction,
+    ) -> Result<eth::Gas, Error> {
+        // Our settlement contract will fail if the receiver is a smart contract.
+        // Because of this, if the receiver is a smart contract and we try to
+        // estimate the access list, the access list estimation will also fail.
+        //
+        // This failure happens is because the Ethereum protocol sets a hard gas limit
+        // on transferring ETH into a smart contract, which some contracts exceed unless
+        // the access list is already specified.
+        //
+        // The solution is to do access list estimation in two steps: first, simulate
+        // moving 1 wei into every smart contract to get a partial access list,
+        // and then simulate the full access list, passing the partial access
+        // list into the simulation. This way the settlement contract does not
+        // fail, and hence the full access list estimation also does not fail.
+        let partial_access_lists = try_join_all(
+            self.user_trades()
+                .map(|trade| async { self.partial_access_list(eth, simulator, trade).await }),
+        )
+        .await?;
+        let partial_access_list = partial_access_lists
+            .into_iter()
+            .fold(eth::AccessList::default(), |acc, list| acc.merge(list));
+        let settlement = Settlement::encode(eth, auction, self).await?.tx();
+        let access_list = simulator
+            .access_list(&settlement, partial_access_list)
+            .await?;
+        simulator
+            .gas(&settlement, &access_list)
+            .await
+            .map_err(Into::into)
+    }
+
+    async fn partial_access_list(
+        &self,
+        eth: &Ethereum,
+        simulator: &Simulator,
+        trade: &trade::Fulfillment,
+    ) -> Result<eth::AccessList, Error> {
+        if !trade.order.buys_eth() || !trade.order.pays_to_contract(eth).await? {
+            return Ok(Default::default());
+        }
+        let tx = eth::Tx {
+            from: self.solver.account(),
+            to: trade.order.receiver(),
+            value: 1.into(),
+            input: Vec::new(),
+        };
+        simulator
+            .access_list(&tx, eth::AccessList::default())
+            .await
+            .map_err(Into::into)
+    }
```

Comment:
Is there a specific reason for access list not being a part of the `eth::Tx`? At the end it always end up as a part of the transaction, and we would avoid forwarding access list as a separate parameter to `simulator.access_list` function etc.

Reply:
I think conceptually (at least since EIP-2930), access lists are part of transactions.